src/driver/odoc_unit.ml

```diff
@@ -0,0 +1,231 @@
+type pkg_args = {
+  pages : (string * Fpath.t) list;
+  libs : (string * Fpath.t) list;
+}
+
+type index = {
+  pkg_args : pkg_args;
+  output_file : Fpath.t;
+  json : bool;
+  search_dir : Fpath.t;
+}
+
+type 'a unit = {
+  parent_id : Odoc.id;
+  odoc_dir : Fpath.t;
+  input_file : Fpath.t;
+  output_dir : Fpath.t;
+  odoc_file : Fpath.t;
+  odocl_file : Fpath.t;
+  pkg_args : pkg_args;
+  pkgname : string;
+  include_dirs : Fpath.t list;
+  index : index;
+  kind : 'a;
+}
+
+type intf_extra = { hidden : bool; hash : string; deps : intf unit list }
+and intf = [ `Intf of intf_extra ]
+
+type impl_extra = { src_id : Odoc.id; src_path : Fpath.t }
+type impl = [ `Impl of impl_extra ]
+
+type mld = [ `Mld ]
+
+type t = [ impl | intf | mld ] unit
+
+let of_packages ~output_dir ~linked_dir ~index_dir (pkgs : Packages.t list) :
+    t list =
+  let linked_dir =
+    match linked_dir with None -> output_dir | Some dir -> dir
+  in
+  let index_dir = match index_dir with None -> output_dir | Some dir -> dir in
+  (* This isn't a hashtable, but a table of hashes! Yay! *)
+  let hashtable =
+    let open Packages in
+    let h = Util.StringMap.empty in
+    List.fold_left
+      (fun h pkg ->
+        List.fold_left
+          (fun h lib ->
+            List.fold_left
+              (fun h mod_ ->
+                Util.StringMap.add mod_.m_intf.mif_hash
+                  (pkg, lib.lib_name, mod_) h)
+              h lib.modules)
+          h pkg.libraries)
+      h pkgs
+  in
+  (* This one is a hashtable *)
+  let cache = Hashtbl.create 10 in
+  let pkg_args_of pkg : pkg_args =
+    let pages =
+      [
+        (pkg.Packages.name, Fpath.(output_dir // pkg.Packages.pkg_dir / "doc"));
+      ]
+    in
+    let libs =
+      List.map
+        (fun lib ->
+          ( lib.Packages.lib_name,
+            Fpath.(output_dir // pkg.Packages.pkg_dir / "lib" / lib.lib_name) ))
+        pkg.libraries
+    in
+    { pages; libs }
+  in
+  let pkg_args : pkg_args =
+    let pages, libs =
+      List.fold_left
+        (fun (all_pages, all_libs) pkg ->
+          let { pages; libs } = pkg_args_of pkg in
+          (pages :: all_pages, libs :: all_libs))
+        ([], []) pkgs
+    in
+    let pages = List.concat pages in
+    let libs = List.concat libs in
+    { pages; libs }
+  in
+  let index_of pkg =
+    let pkg_args = pkg_args_of pkg in
+    let output_file = Fpath.(index_dir / pkg.name / Odoc.index_filename) in
+    { pkg_args; output_file; json = false; search_dir = pkg.pkg_dir }
+  in
+  let rec of_intf hidden pkg libname (intf : Packages.intf) : intf unit =
```

Comment:
There's too much duplicated code in these 3 functions. The code that differ between them is surprising small: `kind`, `odoc_file` and `odocl_file` differ between `of_intf` and `of_impl`. `of_mld` is very similar too.

Reply:
Unfortunately, the types makes it a bit difficult to factor all of those into a single `make_unit` function (`of_intf` returns an `intf unit` while `of_impl` returns an `impl unit`). I'll try!

Reply:
In the end the `make_unit` function returns a "`unit/2 unit`" (shadowing the unit type is not great...) and the extra part is added outside of `make_unit`.